### DISCUSSION TITLE
rp2: re.match with complicated expression crashes MicroPython requiring hard reset

### DISCUSSION THREAD

**Discussion by pdg137** [micropython/micropython#13319] [RP2040 / Pico]
On a Pico with the latest firmware, the follow regex crashes the system completely, requiring a reset / power cycle to recover:

```python
MicroPython v1.22.0 on 2023-12-27; Raspberry Pi Pico with RP2040

Type "help()" for more information.

>>> import re
re.match('[ab]*[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]*$','cbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbb')

```

Of course regular expressions can be arbitrarily complicated, but it seems like there must be a mistake if the board is becoming unresponsive.

On CPython the same command returns a result in a second or two:

```python
Python 3.10.11 (C:\Program Files (x86)\Thonny\python.exe)
>>> import re
re.match('[ab]*[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]*$','cbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbb')
<re.Match object; span=(0, 74), match='cbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbb>
>>> 
```

**Comment by robert-hh** [comment]
That example succeeds here, but it takes a while. Times:
- MIMXRT1176: 30 seconds
- Teensy 4.x: 50 seconds
- PYBD SF6: 164 seconds
- ESP32C3: 310 seconds
- Pyboard V1.1: 331 seconds
- ESP32S3: 406 seconds
- SAMD51: 420 seconds
- ESP32 WROOM: 477 seconds
- RP2: 499 seconds
- ESP32 WROVER with SPIRAM: 570 seconds

All boards at their default clock frequencies. And no, it cannot be stopped with Ctrl-C.
```
=== import re
=== re.match('[ab]*[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]+[cb][cb][cb][cb][ab]+[cb][ab]+[cb][ab]+[cb][ab]*$','cbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbbbcbbbbbbabbabbb')
=== 
=== 
<match num=1>
```
Some other candidates:
- 12 year old i5 PC: 2.7 seconds
- RPI3: 21.4 seconds
- SAMD21: 1321 seconds
- w600: 2894 seconds
- ESP8266: fails almost immediately with reset cause 2 (watchdog?)

So I do not think it's bug, but some of the times are quite extreme.

**Comment by pdg137** [comment]
Wow, I didn't think of letting it run to completion. So is it just standard that library calls can't be interrupted?

**Comment by robert-hh** [comment]
> So is it just standard that library calls can't be interrupted?

It depends. If it's expected that calls can take long, like e.g. uart.read(), they are made interruptible. For re.match() this seems not to be expected.
